sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1162 @@
+
+Node is a class that encapsulates tasks like SYCL kernel functions, device
+memory allocations/frees, or host tasks for deferred execution. A graph has to
+be created first, the structure of a graph is defined second by adding nodes and
+edges.
+
+The `node` class provides the {crs}[common reference semantics].
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+  class node {};
+}
+----
+
+==== Depends-On Property
+
+The API for explicitly adding nodes to a `command_graph` includes a
+`property_list` parameter. This extension defines the `depends_on` property to
+be passed here. `depends_on` defines any `node` objects for the created node to
+be dependent on, and therefore form an edge with. These nodes are in addition to
+the dependent nodes identified from the command-group requisites of the created
+node.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental::property::node
+class depends_on {
+  public:
+    template<typename... NodeTN>
+    depends_on(NodeTN... nodes);
+};
+}
+----
+
+=== Graph
+
+This extension adds a new `command_graph` object which follows the
+{crs}[common reference semantics] of other SYCL runtime objects.
+
+A `command_graph` represents a directed acyclic graph of nodes, where each node
+represents a single command or a sub-graph. The execution of a graph completes
+when all of its nodes have completed.
+
+A `command_graph` is built up by either recording queue submissions or
+explicitly adding nodes, then once the user is happy that the graph is complete,
+the graph instance is finalized into an executable variant which can have no
+more nodes added to it. Finalization may be a computationally expensive
+operation as the runtime is able to perform optimizations based on the graph
+structure. After finalization the graph can be submitted for execution on a
+queue one or more times with reduced overhead.
+
+==== Graph State
+
+An instance of a `command_graph` object can be in one of two states:
+
+* **Modifiable** - Graph is under construction and new nodes may be added to it.
+* **Executable** - Graph topology is fixed after finalization and graph is ready to
+  be submitted for execution.
+
+A `command_graph` object is constructed in the _recording_ state and is made
+_executable_ by the user invoking `command_graph::finalize()` to create a
+new executable instance of the graph. An executable graph cannot be converted
+to a modifiable graph. After finalizing a graph in the modifiable state it is
+valid for a user to add additional nodes and finalize again to create subsequent
+executable graphs. The state of a `command_graph` object is made explicit by
+templating on state to make the class strongly typed, with the default template
+argument being `graph_state::modifiable` to reduce code verbosity on
+construction.
+
+.Graph State Diagram
+image::images/sycl_ext_oneapi_graph/command_graph-state.svg[]
```

Comment:
I'd recommend replacing these with a Mermaid diagram, since you can embed and edit them on GitHub.

This code:
```
graph LR
    Modifiable -->|Finalize| Executable
```

Renders as:
```mermaid
graph LR
    Modifiable -->|Finalize| Executable
```

Same goes for the other SVG.

Reply:
I hadn't heard of Mermaid diagrams before, very cool 🙂  Update to that here - https://github.com/reble/llvm/pull/51